File: src/dk/xam/jhostty/ZmxSession.java
package dk.xam.jhostty;

import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

public record ZmxSession(String name, int pid, int clients, String startDir, String cwd, String cmd, boolean ended, int exitCode) {

    public String displayLabel() {
        var status = ended ? " \u2718" : (clients > 0 ? " (" + clients + ")" : "");
        return friendlyName() + status;
    }

    public String friendlyName() {
        var displayName = name;
        if (isGeneratedName()) {
            var dir = cwd != null && !cwd.isBlank() ? cwd : startDir;
            if (dir != null && !dir.isBlank()) {
                var p = Path.of(dir);
                var home = Path.of(System.getProperty("user.home", ""));
                if (p.equals(home)) {
                    displayName = "~";
                } else if (p.startsWith(home) && p.getNameCount() > home.getNameCount()) {
                    var rel = home.relativize(p);
                    var count = rel.getNameCount();
                    displayName = count >= 2
                            ? rel.getName(count - 2) + "/" + rel.getName(count - 1)
                            : rel.getFileName().toString();
                } else {
                    displayName = p.getFileName().toString();
                }
            }
        }
        return displayName;
    }

    public boolean isGeneratedName() {
        return name.matches(".*[0-9a-f]{8}-[0-9a-f]{4}-.*");
    }

    // --- Parsing ---

    public static List<ZmxSession> parseZmxList(String output) {
        if (output == null || output.isBlank()) return List.of();
        var result = new ArrayList<ZmxSession>();
        for (var line : output.split("\n")) {
            line = line.trim();
            if (line.startsWith("\u2192 ")) line = line.substring(2).trim();
            if (line.isBlank()) continue;
            var fields = new java.util.HashMap<String, String>();
            for (var part : line.split("\t")) {
                part = part.trim();
                var eq = part.indexOf('=');
                if (eq > 0) {
                    fields.put(part.substring(0, eq).trim(), part.substring(eq + 1).trim());
                }
            }
            var name = fields.getOrDefault("name", "");
            if (name.isBlank()) continue;
            result.add(new ZmxSession(
                    name,
                    parseInt(fields.getOrDefault("pid", "0")),
                    parseInt(fields.getOrDefault("clients", "0")),
                    fields.getOrDefault("start_dir", ""),
                    fields.getOrDefault("cwd", ""),
                    fields.getOrDefault("cmd", ""),
                    fields.containsKey("ended"),
                    parseInt(fields.getOrDefault("exit_code", "0"))
            ));
        }
        return result;
    }

    private static int parseInt(String s) {
        try { return Integer.parseInt(s); } catch (NumberFormatException _) { return 0; }
    }
}

File: ios/Ghostly/Models/ZmxSession.swift
import Foundation

/// A persistent `zmx` session living on a remote host.
///
/// This is a faithful Swift port of `dk.xam.jhostty.ZmxSession` from the
/// desktop jhostty app, so the two clients agree on how `zmx list` output is
/// parsed and how sessions are named for display.
///
/// `zmx list` prints one session per line as tab-separated `key=value` pairs,
/// with the currently-attached session prefixed by `→ `. Example:
///
/// ```
/// → name=dotfiles	pid=4321	clients=1	start_dir=/home/max	cwd=/home/max/.config	cmd=zsh
///   name=b1f2c3d4-...	pid=4400	clients=0	start_dir=/srv	cwd=/srv	cmd=zsh	ended	exit_code=0
/// ```
struct ZmxSession: Identifiable, Hashable {
    let name: String
    let pid: Int
    let clients: Int
    let startDir: String
    let cwd: String
    let cmd: String
    let ended: Bool
    let exitCode: Int

    var id: String { name }

    /// Whether any client is currently attached to this session.
    var isAttached: Bool { clients > 0 }

    /// Short status suffix used in list rows, e.g. `dotfiles (1)` or `srv ✘`.
    var displayLabel: String {
        let status = ended ? " ✘" : (clients > 0 ? " (\(clients))" : "")
        return friendlyName + status
    }

    /// zmx auto-generates UUID-style names for unnamed sessions. When that
    /// happens we present the working directory instead, mirroring the desktop
    /// app's behaviour (last two path components, `~` for home).
    var friendlyName: String {
        guard isGeneratedName else { return name }
        let dir = !cwd.isEmpty ? cwd : startDir
        guard !dir.isEmpty else { return name }

        let home = NSHomeDirectory()
        let path = (dir as NSString).standardizingPath
        if path == home { return "~" }
        if !home.isEmpty, path.hasPrefix(home + "/") {
            let rel = String(path.dropFirst(home.count + 1))
            let parts = rel.split(separator: "/")
            if parts.count >= 2 {
                return "\(parts[parts.count - 2])/\(parts[parts.count - 1])"
            }
            return parts.last.map(String.init) ?? name
        }
        return (path as NSString).lastPathComponent
    }

    /// Matches the UUID embedded in zmx's auto-generated session names.
    var isGeneratedName: Bool {
        name.range(of: "[0-9a-f]{8}-[0-9a-f]{4}-", options: .regularExpression) != nil
    }

    // MARK: - Parsing

    /// Parse the raw stdout of `zmx list` into a list of sessions.
    static func parseList(_ output: String) -> [ZmxSession] {
        guard !output.isEmpty else { return [] }
        var result: [ZmxSession] = []

        for rawLine in output.split(whereSeparator: \.isNewline) {
            var line = rawLine.trimmingCharacters(in: .whitespaces)
            if line.hasPrefix("→ ") { line = String(line.dropFirst(2)).trimmingCharacters(in: .whitespaces) }
            if line.isEmpty { continue }

            var fields: [String: String] = [:]
            for part in line.split(separator: "\t") {
                let piece = part.trimmingCharacters(in: .whitespaces)
                guard let eq = piece.firstIndex(of: "="), eq > piece.startIndex else { continue }
                let key = String(piece[piece.startIndex..<eq]).trimmingCharacters(in: .whitespaces)
                let value = String(piece[piece.index(after: eq)...]).trimmingCharacters(in: .whitespaces)
                fields[key] = value
            }

            let name = fields["name"] ?? ""
            if name.isEmpty { continue }

            result.append(
                ZmxSession(
                    name: name,
                    pid: Int(fields["pid"] ?? "0") ?? 0,
                    clients: Int(fields["clients"] ?? "0") ?? 0,
                    startDir: fields["start_dir"] ?? "",
                    cwd: fields["cwd"] ?? "",
                    cmd: fields["cmd"] ?? "",
                    ended: fields.keys.contains("ended"),
                    exitCode: Int(fields["exit_code"] ?? "0") ?? 0
                )
            )
        }
        return result
    }
}
